samples/net/latmon/src/main.c

```diff
@@ -0,0 +1,263 @@
+}
+
+static int configure_measurement_hardware(void)
+{
+	static struct gpio_callback gpio_cb = { };
+	int ret = 0;
+
+	if (!gpio_is_ready_dt(&pulse) || !gpio_is_ready_dt(&ack)) {
+		LOG_ERR("GPIO device not ready");
+		return -ENODEV;
+	}
+
+	ret = gpio_pin_configure_dt(&pulse, GPIO_OUTPUT_HIGH);
+	if (ret < 0) {
+		LOG_ERR("failed configuring pulse pin");
+		return ret;
+	}
+
+	ret = gpio_pin_configure_dt(&ack, GPIO_INPUT);
+	if (ret < 0) {
+		LOG_ERR("failed configuring ack pin");
+		return ret;
+	}
+
+#if defined(LATMON_LOOPBACK_DBG)
+	/*
+	 * Connect GPIO pins in loopback mode for validation (tx to ack)
+	 * On FRDM_K64F, Latmus will show around 3.2 usec of latency.
+	 */
+	ret = gpio_pin_interrupt_configure_dt(&ack, GPIO_INT_EDGE_FALLING);
+#else
+	ret = gpio_pin_interrupt_configure_dt(&ack, GPIO_INT_EDGE_RISING);
+#endif
+	if (ret < 0) {
+		LOG_ERR("failed configuring ack pin interrupt");
+		return ret;
+	}
+
+	gpio_init_callback(&gpio_cb, gpio_ack_handler, BIT(ack.pin));
+
+	ret = gpio_add_callback_dt(&ack, &gpio_cb);
+	if (ret < 0) {
+		LOG_ERR("failed adding ack pin callback");
+		return ret;
+	}
+
+	return ret;
+}
+
+static void blink(void*, void*, void*)
+{
+	const struct gpio_dt_spec led_run =
+		GPIO_DT_SPEC_GET_OR(DT_ALIAS(led0), gpios, {0});
+	const struct gpio_dt_spec led_wait =
+		GPIO_DT_SPEC_GET_OR(DT_ALIAS(led1), gpios, {0});
+	const struct gpio_dt_spec led_dhcp =
+		GPIO_DT_SPEC_GET_OR(DT_ALIAS(led2), gpios, {0});
+	const struct gpio_dt_spec *led = &led_dhcp, *tmp = NULL;
+	uint32_t period = LED_DHCP_PERIOD;
+
+	if (gpio_is_ready_dt(&led_run)) {
+		gpio_pin_configure_dt(&led_run, GPIO_OUTPUT_INACTIVE);
+	}
+
+	if (gpio_is_ready_dt(&led_wait)) {
+		gpio_pin_configure_dt(&led_wait, GPIO_OUTPUT_INACTIVE);
+	}
+
+	if (gpio_is_ready_dt(&led_dhcp)) {
+		gpio_pin_configure_dt(&led_dhcp, GPIO_OUTPUT_INACTIVE);
+	}
+
+	for (;;) {
+		k_usleep(period);
+		if (dhcp_done) {
+			led = net_latmon_running() ? &led_run : &led_wait;
+		}
+
+		if (tmp && led != tmp) {
+			gpio_pin_set_dt(tmp, 0);
+		}
+
+		if (!gpio_is_ready_dt(led)) {
+			continue;
+		}
+
+		if (led == &led_wait) {
+			period = LED_WAIT_PERIOD;
+		}
+
+		if (led == &led_run) {
+			period = LED_RUN_PERIOD;
+		}
+
+		gpio_pin_toggle_dt(led);
+		tmp = led;
+	}
+	gpio_pin_set_dt(led, 0);
+}
+
+static k_tid_t start_led_blinking_thread(struct k_thread *blink_thread,
+					k_thread_entry_t blink_thread_func)
+{
+	return k_thread_create(blink_thread, blink_stack, BLINK_STACK_SIZE,
+				(k_thread_entry_t)blink_thread_func,
+				NULL, NULL, NULL,
+				BLINK_THREAD_PRIORITY, 0, K_NO_WAIT);
+}
+
+/* Raw ticks */
+#define CALCULATE_DELTA(ack, pulse) \
+((ack) < (pulse) ? \
+(~(pulse) + 1 + (ack)) : ((ack) - (pulse)))
+
+static int measure_latency_cycles(uint32_t *delta)
+{
+	unsigned int key = 0;
+	uint32_t tx = 0;
+	uint32_t rx = 0;
+	int ret = 0;
+
+	/* Remove spurious events */
+	k_sem_reset(&ack_event);
+
+	/* Generate a falling edge pulse to the DUT */
+	key = irq_lock();
```

Comment:
applications should use kernel primitives, `irq_lock()` is just there for compatibility reasons, instead, `struct k_spinlock` should be used for critical sections.

Reply:
ok